splitgpg2/__init__.py

```diff
@@ -0,0 +1,804 @@
+
+    def client_write(self, data):
+        self.log_io('C <<<', data)
+        self.client_transport.write(data)
+
+    def eof_received(self) -> Optional[bool]:
+        # close connection to the real gpg agent too
+        if self.agent_transport is not None:
+            self.agent_transport.write_eof()
+        # automatically close the transport
+        return False
+
+    def data_received(self, untrusted_data: bytes) -> None:
+        # pylint: disable=arguments-differ
+        if len(self.untrusted_input_buffer) > ASSUAN_LINELENGTH*2:
+            self.log.error('Too much data received, dropping')
+
+        self.log_io('C >>>', untrusted_data)
+
+        self.untrusted_input_buffer.extend(untrusted_data)
+        if self.state == ServerState.agent_response:
+            # now is an agent turn, only buffer the data, but don't act on it
+            return
+
+        while b'\n' in self.untrusted_input_buffer:
+            # there is a full command buffered
+            untrusted_line, untrusted_rest = \
+                self.untrusted_input_buffer.split(b'\n', 1)
+            self.untrusted_input_buffer = untrusted_rest
+            # convert from bytesarray to bytes to be hashable
+            untrusted_line = bytes(untrusted_line)
+            if self.state == ServerState.client_command:
+                asyncio.ensure_future(self.handle_command(
+                    untrusted_line=untrusted_line))
+            elif self.state == ServerState.client_inquire:
+                asyncio.ensure_future(self.handle_inquire(
+                    untrusted_line=untrusted_line))
+
+    async def send_inquire(self, inquire, inquire_commands):
+        if self.state != ServerState.agent_response:
+            raise ProtocolError(
+                'invalid server state: INQUIRE not originating from gpg-agent')
+        loop = asyncio.get_event_loop()
+        self.inquire_complete = loop.create_future()
+        self.inquire_commands = inquire_commands
+        self.state = ServerState.client_inquire
+        self.client_write(b'INQUIRE ' + inquire + b'\n')
+        await self.inquire_complete
+        self.inquire_commands = {}
+        self.state = ServerState.agent_response
+
+    def fake_respond(self, response):
+        self.client_write(response + b'\n')
+
+    @staticmethod
+    def verify_keygrip_arguments(min_count, max_count, untrusted_args):
+        args_regex = re.compile(
+            rb'[0-9A-F]{40}( [0-9A-F]{40}){%d,%d}' % (min_count-1, max_count-1))
+
+        if args_regex.fullmatch(untrusted_args) is None:
+            raise Filtered
+        return untrusted_args
+
+    def sanitize_key_desc(self, untrusted_args: bytes) -> bytes:
+        untrusted_args = untrusted_args.replace(b'+', b' ')
+        untrusted_args = re.sub(
+            rb'%[0-9A-F]{2}',
+            lambda m: bytes([int(m.group(0)[1:].decode('ascii'), 16)]),
+            untrusted_args,
+            flags=re.IGNORECASE
```

Comment:
`INGORECASE` derivatives from the Ruby code. Why?

If this is needed please split language conversion and functional change.